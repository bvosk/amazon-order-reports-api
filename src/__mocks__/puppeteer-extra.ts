import { Browser, Page, Response } from 'puppeteer';

const response = ({
  ok: jest.fn()
} as unknown) as Response;

const page = ({
  $: jest.fn(),
  $$eval: jest.fn(),
  click: jest.fn(),
  goto: jest.fn(),
  type: jest.fn(),
  url: jest.fn(),
  waitForNavigation: jest.fn()
} as unknown) as Page;

const browser = ({
  close: jest.fn(),
  newPage: jest.fn()
} as unknown) as Browser;

export const mocks = {
  browser,
  page,
  response
};

export default {
  launch: jest.fn(),
  use: jest.fn()
};
