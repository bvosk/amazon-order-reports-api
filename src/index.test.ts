import mockdate from 'mockdate';
import { ElementHandle } from 'puppeteer';
import puppeteer from 'puppeteer-extra';
import { mocked } from 'ts-jest/utils';
import AmazonScraper, { TimePeriod } from './index';
import { mocks } from './__mocks__/puppeteer-extra';

describe('AmazonScraper', () => {
  beforeEach(() => {
    mocked(puppeteer.launch).mockResolvedValue(mocks.browser);
    mocked(mocks.browser.newPage).mockResolvedValue(mocks.page);
    mocked(mocks.page.goto).mockResolvedValue(mocks.response);
    mocked(mocks.page.$).mockResolvedValue(({} as unknown) as ElementHandle<Element>);
    mocked(mocks.page.url).mockReturnValue('');
    mocked(mocks.response.ok).mockReturnValue(true);

    mockdate.set('2020-01-01T00:00:00.000Z');
  });

  afterEach(() => {
    jest.resetAllMocks();
    mockdate.reset();
  });

  const mockUrls = (...urls: Array<string>) => {
    mocked(mocks.page.url).mockImplementation(() => urls.shift() ?? '');
  };

  describe('start', () => {
    it('should launch puppeteer with given options', async () => {
      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        puppeteerOpts: {
          defaultViewport: {
            height: 600,
            width: 800
          }
        }
      });

      await scraper.start();

      expect(mocked(puppeteer.launch)).toBeCalledWith({
        defaultViewport: {
          height: 600,
          width: 800
        }
      });
    });
  });

  describe('stop', () => {
    it('should close puppeteer', async () => {
      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        puppeteerOpts: {
          defaultViewport: {
            height: 600,
            width: 800
          }
        }
      });

      await scraper.start();
      await scraper.stop();

      expect(mocked(mocks.browser.close)).toBeCalled();
    });
  });

  describe('login', () => {
    it('should log in with given credentials', async () => {
      mockUrls('https://www.amazon.com/ap/signin');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await scraper.getOrders().next();

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith(
        'input[name=email]',
        'testuser@example.com'
      );

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith('input[name=password]', 'test123');
      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('input[name=rememberMe]');
    });

    it('should throw error if OTP code is required, but no OTP option is provided', async () => {
      mockUrls('https://www.amazon.com/ap/signin', 'https://www.amazon.com/ap/mfa');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await expect(scraper.getOrders().next()).rejects.toThrow(
        '2FA code requested, but neither otpSecret nor otpFn were provided'
      );
    });

    it('should login using generated OTP code if otpSecret is provided', async () => {
      mockUrls('https://www.amazon.com/ap/signin', 'https://www.amazon.com/ap/mfa');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        otpSecret: 'GVYX4RKDFYXXMNDJ'
      });

      await scraper.getOrders().next();

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith('input[name=otpCode]', '930899');
      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('input[name=rememberDevice]');
    });

    it('should login using OTP code provided by otpFn', async () => {
      mockUrls('https://www.amazon.com/ap/signin', 'https://www.amazon.com/ap/mfa');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        otpFn: async () => '123456'
      });

      await scraper.getOrders().next();

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith('input[name=otpCode]', '123456');
      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('input[name=rememberDevice]');
    });

    it('should skip account fix-up page during login', async () => {
      mockUrls(
        'https://www.amazon.com/ap/signin',
        'https://www.amazon.com/ap/accountfixup',
        'https://www.amazon.com/ap/accountfixup'
      );

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await scraper.getOrders().next();

      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('a[id*=skip]');
    });

    it('should throw error if sign-out assertion fails', async () => {
      mockUrls('https://www.amazon.com/ap/signin');
      mocked(mocks.page.$).mockResolvedValue(null);

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await expect(scraper.getOrders().next()).rejects.toThrow('No sign-out link detected');
    });

    it('should throw error on bad response', async () => {
      mockUrls('https://www.amazon.com/ap/signin');
      mocked(mocks.response.ok).mockReturnValue(false);

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await expect(scraper.getOrders().next()).rejects.toThrow('Failed request');
    });
  });

  describe('getOrders', () => {
    let scraper: AmazonScraper;
    let pages: {
      [index: string]: Array<{ [selector: string]: Array<{ [attr: string]: string }> }>;
    };

    beforeEach(() => {
      scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      pages = {
        '0': [
          {
            '.value': [
              { textContent: 'February 29, 2020' },
              { textContent: '$13.85' },
              { textContent: 'Scrandleton Sprunt' },
              { textContent: '123-6543210-0123456' }
            ]
          },
          {
            '.value': [
              { textContent: 'January 20, 2020' },
              { textContent: '$107.95' },
              { textContent: '098-1234567-7654321' }
            ]
          },
          {
            '.value': [
              { textContent: 'January 1, 2020' },
              { textContent: '$10' },
              { textContent: 'Scrandleton Sprunt' },
              { textContent: '876-4567890-0987654' }
            ]
          }
        ],
        '10': [
          {
            '.value': [
              { textContent: 'February 29, 2019' },
              { textContent: '$13.85' },
              { textContent: 'Scrandleton Sprunt' },
              { textContent: '223-6543210-0123456' }
            ]
          },
          {
            '.value': [
              { textContent: 'January 20, 2019' },
              { textContent: '$107.95' },
              { textContent: '198-1234567-7654321' }
            ]
          },
          {
            '.value': [
              { textContent: 'January 1, 2019' },
              { textContent: '$10' },
              { textContent: 'Scrandleton Sprunt' },
              { textContent: 'D01-4567890-0987654' }
            ]
          }
        ]
      };

      let orderInfos: typeof pages[0];

      mocked(mocks.page.goto).mockImplementation(async (url: string) => {
        const { startIndex } = url.match(/startIndex=(?<startIndex>\d+)/)?.groups ?? {};
        if (startIndex) {
          orderInfos = pages[startIndex] ?? [];
        } else {
          orderInfos = [];
        }

        return mocks.response;
      });

      mocked(mocks.page.$$eval).mockImplementation(
        async <R, X1, X2>(
          selector: string,
          fn: (els: Array<Element>, x1: X1, x2: X2) => R | Promise<R>,
          x1: X1,
          x2: X2
        ) => {
          if (selector !== '.order-info') {
            return [];
          }

          return fn(
            orderInfos.map(
              (orderInfo) =>
                (({
                  querySelectorAll: (selector: string) => orderInfo[selector]
                } as unknown) as Element)
            ),
            x1,
            x2
          );
        }
      );
    });

    it('should return order info', async () => {
      const results = [];
      for await (const order of scraper.getOrders()) {
        results.push(order);
      }

      expect(results).toEqual([
        {
          id: '123-6543210-0123456',
          date: 'February 29, 2020',
          shipTo: 'Scrandleton Sprunt',
          total: '$13.85'
        },
        { id: '098-1234567-7654321', date: 'January 20, 2020', total: '$107.95' },
        {
          id: '876-4567890-0987654',
          date: 'January 1, 2020',
          shipTo: 'Scrandleton Sprunt',
          total: '$10'
        },
        {
          date: 'February 29, 2019',
          id: '223-6543210-0123456',
          shipTo: 'Scrandleton Sprunt',
          total: '$13.85'
        },
        {
          date: 'January 20, 2019',
          id: '198-1234567-7654321',
          total: '$107.95'
        },
        {
          date: 'January 1, 2019',
          id: 'D01-4567890-0987654',
          shipTo: 'Scrandleton Sprunt',
          total: '$10'
        }
      ]);
    });

    it('should throw error on failure to parse order ID', async () => {
      pages[0].push({
        '.value': [
          { textContent: 'December 25, 2019' },
          { textContent: '$95.00' },
          { textContent: 'weird content' }
        ]
      });

      await expect(async () => {
        for await (const _ of scraper.getOrders()) {
        }
      }).rejects.toThrow('Failed to parse order ID');
    });

    it('should throw error on too few values', async () => {
      pages[0].push({
        '.value': [{ textContent: 'December 25, 2019' }, { textContent: '$95.00' }]
      });

      await expect(async () => {
        for await (const _ of scraper.getOrders()) {
        }
      }).rejects.toThrow('Order info has fewer values than expected');
    });

    it('should stop paging at limit', async () => {
      const results = [];
      for await (const order of scraper.getOrders({ limit: 10 })) {
        results.push(order);
      }

      expect(results.map((r) => r.id)).toEqual([
        '123-6543210-0123456',
        '098-1234567-7654321',
        '876-4567890-0987654'
      ]);
    });

    it('start from given startIndex', async () => {
      const results = [];
      for await (const order of scraper.getOrders({ startIndex: 10 })) {
        results.push(order);
      }

      expect(results.map((r) => r.id)).toEqual([
        '223-6543210-0123456',
        '198-1234567-7654321',
        'D01-4567890-0987654'
      ]);
    });

    it('should handle years in timePeriod', async () => {
      for await (const _ of scraper.getOrders({ timePeriod: 2020 })) {
      }

      expect(mocked(mocks.page.goto)).toBeCalledWith(
        'https://www.amazon.com/gp/your-account/order-history?orderFilter=year-2020&startIndex=0'
      );
    });

    it('should handle LAST_3_MONTHS in timePeriod', async () => {
      for await (const _ of scraper.getOrders({ timePeriod: TimePeriod.LAST_3_MONTHS })) {
      }

      expect(mocked(mocks.page.goto)).toBeCalledWith(
        'https://www.amazon.com/gp/your-account/order-history?orderFilter=months-3&startIndex=0'
      );
    });
  });
});
