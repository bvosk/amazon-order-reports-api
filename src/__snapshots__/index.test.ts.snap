// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmazonScraper getItems should return item for each row in report 1`] = `
Array [
  Object {
    "asinIsbn": "B06Y5XG33Q",
    "buyerName": "Test Man",
    "carrierNameTrackingNumber": "USPS(9300120111405739275629)",
    "category": "HEALTH_PERSONAL_CARE",
    "condition": "new",
    "currency": "USD",
    "itemSubtotal": 21.75,
    "itemSubtotalTax": 0,
    "itemTotal": 21.75,
    "listPricePerUnit": 0,
    "orderDate": 2020-01-03T08:00:00.000Z,
    "orderId": "123-7654321-1234567",
    "orderStatus": "Shipped",
    "orderingCustomerEmail": "email@example.com",
    "paymentInstrumentType": "Visa - 1234",
    "purchasePricePerUnit": 21.75,
    "quantity": 1,
    "seller": "This Is A Seller",
    "shipmentDate": 2020-01-06T08:00:00.000Z,
    "shippingAddressCity": "Washington",
    "shippingAddressName": "Test Man",
    "shippingAddressState": "DC",
    "shippingAddressStreet1": "1600 Pennsylvania Avenue NW",
    "shippingAddressZip": "20500",
    "title": "Some Product",
    "unspscCode": "38462846",
    "website": "Amazon.com",
  },
  Object {
    "asinIsbn": "B07X9FXSPL",
    "buyerName": "Test Man",
    "carrierNameTrackingNumber": "AMZN_US(TBA014759273649)",
    "category": "LIGHT_BULB",
    "condition": "new",
    "currency": "USD",
    "itemSubtotal": 59.94,
    "itemSubtotalTax": 0,
    "itemTotal": 59.94,
    "listPricePerUnit": 24.99,
    "orderDate": 2020-04-20T07:00:00.000Z,
    "orderId": "321-9876543-3456789",
    "orderStatus": "Shipped",
    "orderingCustomerEmail": "email@example.com",
    "paymentInstrumentType": "Visa - 4321",
    "purchasePricePerUnit": 19.98,
    "quantity": 3,
    "seller": "Amazon.com",
    "shipmentDate": 2020-04-21T07:00:00.000Z,
    "shippingAddressCity": "Washington",
    "shippingAddressName": "Test Man",
    "shippingAddressState": "DC",
    "shippingAddressStreet1": "1600 Pennsylvania Avenue NW",
    "shippingAddressZip": "20500",
    "title": "Another Product",
    "unspscCode": "65927561",
    "website": "Amazon.com",
  },
  Object {
    "asinIsbn": "B0159OP2FF",
    "buyerName": "Test Man",
    "carrierNameTrackingNumber": "AMZN_US(TBA023749506528)",
    "category": "STANDALONE_STREAMING_MEDIA_PLAYER",
    "condition": "new",
    "currency": "USD",
    "itemSubtotal": 69.69,
    "itemSubtotalTax": 0,
    "itemTotal": 69.69,
    "listPricePerUnit": 69.69,
    "orderDate": 2020-06-09T07:00:00.000Z,
    "orderId": "987-3456789-7654321",
    "orderStatus": "Shipped",
    "orderingCustomerEmail": "email@example.com",
    "paymentInstrumentType": "Gift Certificate/Card and Visa - 5656",
    "purchasePricePerUnit": 69.69,
    "quantity": 1,
    "seller": "Amazon.com",
    "shipmentDate": 2020-06-10T07:00:00.000Z,
    "shippingAddressCity": "Washington",
    "shippingAddressName": "Test Man",
    "shippingAddressState": "DC",
    "shippingAddressStreet1": "1600 Pennsylvania Avenue NW",
    "shippingAddressZip": "20500",
    "title": "Yet Another Product",
    "unspscCode": "37462840",
    "website": "Amazon.com",
  },
]
`;

exports[`AmazonScraper getRefunds should return refund for each row in report 1`] = `
Array [
  Object {
    "asinIsbn": "B07USNDYEK",
    "buyerName": "Test Man",
    "category": "COAT",
    "orderDate": 2020-01-14T08:00:00.000Z,
    "orderId": "113-5757575-3939393",
    "quantity": 1,
    "refundAmount": 75.98,
    "refundCondition": "Completed",
    "refundDate": 2020-01-25T08:00:00.000Z,
    "refundReason": "Customer Return",
    "refundTaxAmount": 0,
    "seller": "A Seller",
    "title": "Some kind of coat",
    "website": "Amazon.com",
  },
  Object {
    "asinIsbn": "B071UDN9IU",
    "buyerName": "Test Man",
    "category": "SHIRT",
    "orderDate": 2020-02-18T08:00:00.000Z,
    "orderId": "111-1234567-7654321",
    "quantity": 1,
    "refundAmount": 18.99,
    "refundCondition": "Completed",
    "refundDate": 2020-02-26T08:00:00.000Z,
    "refundReason": "Customer Return",
    "refundTaxAmount": 0,
    "seller": "Amazon.com",
    "title": "T-shirt",
    "website": "Amazon.com",
  },
  Object {
    "asinIsbn": "B0793JNURH",
    "buyerName": "Test Man",
    "category": "SHOES",
    "orderDate": 2020-02-28T08:00:00.000Z,
    "orderId": "113-9887789-1234543",
    "quantity": 1,
    "refundAmount": 159.96,
    "refundCondition": "Completed",
    "refundDate": 2020-03-04T08:00:00.000Z,
    "refundReason": "Customer Return",
    "refundTaxAmount": 0,
    "seller": "Amazon.com",
    "title": "A pair of shoes",
    "website": "Amazon.com",
  },
]
`;
